### PR TITLE
Implementation of map tile cache

### PR DESCRIPTION
**Pull request by n76**
See: https://github.com/zdila/gpx-animator/issues/77

Implement a map tile cache to avoid downloading the same files each time
a slightly different rendering is tried. User specifies location of cache
and lifetime of cached tiles. If no cache is specified then revert to old
behavior of downloading tiles each time renderer is run.

This is a partial implementation because only the CLI has been updated to
allow specification of cache parameters. The GUI will need updating to
complete this enhancement.

Signed-off-by: Tod Fitch <tod@fitchfamily.org>

### CODE REVIEW

File: src/main/java/sk/freemap/gpxAnimator/TileCache.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Date;
+import javax.imageio.ImageIO;
+import org.apache.commons.io.FileUtils;
+
+public class TileCache {
+
+    public static BufferedImage getTile( String url, String cachePath, Long cacheTimeLimit ) throws UserException {
```
Comment (McPringle) [inline src/main/java/sk/freemap/gpxAnimator/TileCache.java]: As in the other classes, I would like to see method parameters be marked as `final`. Same with the other methods in this class.

File: src/main/java/sk/freemap/gpxAnimator/TileCache.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+        if (cachePath == null) {
+            //System.out.println("No cache path defined, downloading " + url);
```
Comment (McPringle) [inline src/main/java/sk/freemap/gpxAnimator/TileCache.java]: Please remove this commented code. Same with line 55.

File: src/main/java/sk/freemap/gpxAnimator/TileCache.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        if (mapTile == null) {
+            throw new UserException("could not read tile " + url);
+        }
+        
+        return mapTile;
+    }
+    
+    private static BufferedImage cachedGetTile( String url, String cachePath, Long cacheTimeLimit ) throws UserException {
+        BufferedImage mapTile;
+        String fname = hashName(url);
```
Comment (McPringle) [inline src/main/java/sk/freemap/gpxAnimator/TileCache.java]: I guess the `f` stands for `file`? I would like to rename this variable to `filename`. That removes future misunderstanding of other people working on this code.

File: src/main/java/sk/freemap/gpxAnimator/TileCache.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        String path = cachePath + File.separator + fname;
+        File f = new File(path);
```
Comment (McPringle) [inline src/main/java/sk/freemap/gpxAnimator/TileCache.java]: I don't like one-letter-names for variables... :-)

File: src/main/java/sk/freemap/gpxAnimator/TileCache.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        
+        // Age out old file if needed
+        if (f.isFile()){
+            Date fileDate = new Date(f.lastModified());
+            long msBetweenDates = new Date().getTime() - fileDate.getTime();
+            if ((msBetweenDates/1000) > cacheTimeLimit) {
+                System.out.println("Removing cache entry for " + url);
```
Comment (McPringle) [inline src/main/java/sk/freemap/gpxAnimator/TileCache.java]: The other two "debug" messages are commented out. Should this be removed, too? Same with line 95.

File: src/main/java/sk/freemap/gpxAnimator/TileCache.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        File f = new File(path);
+        
+        // Age out old file if needed
+        if (f.isFile()){
+            Date fileDate = new Date(f.lastModified());
+            long msBetweenDates = new Date().getTime() - fileDate.getTime();
+            if ((msBetweenDates/1000) > cacheTimeLimit) {
+                System.out.println("Removing cache entry for " + url); 
+                if (!f.delete()) {
+                    throw new UserException("failed to delete cache entry for " + url);
```
Comment (McPringle) [inline src/main/java/sk/freemap/gpxAnimator/TileCache.java]: Maybe it is a good idea to add the name of the cache file to the error message because it is different from the URL. With the file name, the user could then check the cache directory and the file for problems.

File: src/main/java/sk/freemap/gpxAnimator/TileCache.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+                } 
+            }
+        }
+        
+        // If map tile is not in cache, then download it.
+        if (!f.isFile()) {
+            System.out.println("Cache miss for:" + path);
+            try {
+                FileUtils.copyURLToFile(new URL(url), f, 5000, 10000);
```
Comment (McPringle) [inline src/main/java/sk/freemap/gpxAnimator/TileCache.java]: How does the HTTP request header look like? In the past when requesting a tile we send "GPX Animator" as the user agent (see line 62). Some APIs require a valid user agent. Otherwise, they might block the requests.

File: src/main/java/sk/freemap/gpxAnimator/TileCache.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+            } catch (final IOException e) {
+                throw new UserException("error downloading tile " + url, e);
+            }
+        }
+        
+        // At this point a current map tile should be in our cache, so
+        // load it for the
```
Comment (McPringle) [inline src/main/java/sk/freemap/gpxAnimator/TileCache.java]: load it for the ??? :-)

File: src/main/java/sk/freemap/gpxAnimator/TileCache.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        if (mapTile == null) {
+            throw new UserException("could not read tile " + url);
+        }
+        
+        return mapTile;
+    }
+    
+    private static String hashName(String url) throws UserException {
+        try {
+            MessageDigest digest = MessageDigest.getInstance("SHA-256");
```
Comment (McPringle) [inline src/main/java/sk/freemap/gpxAnimator/TileCache.java]: Some background maps need a few hundred tiles. You should initialize the message digest on the object level, not method level. You need this always, there is no use case, where this code is never executed.

File: src/main/java/sk/freemap/gpxAnimator/TileCache.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+            return bytesToHex(digest.digest(url.getBytes(StandardCharsets.UTF_8)));
+        } catch (final NoSuchAlgorithmException e) {
+            throw new UserException("error creating hash name " + url, e);
```
Comment (McPringle) [inline src/main/java/sk/freemap/gpxAnimator/TileCache.java]: Not sure, if I like this. There could be an automatic fallback to the no-cache code. If the cache is not working, that should not prevent the application from generating the video.

File: src/main/java/sk/freemap/gpxAnimator/TileCache.java [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+        }
+    }
+    
+    private static String bytesToHex(byte[] hash) {
+        StringBuffer hexString = new StringBuffer();
+        for (int i = 0; i < hash.length; i++) {
+        String hex = Integer.toHexString(0xff & hash[i]);
+        if(hex.length() == 1) hexString.append('0');
```
Comment (McPringle) [inline src/main/java/sk/freemap/gpxAnimator/TileCache.java]: Please correct the indentation. I had to look several times at the code to identify correctly, that line 131 is not executed based on the `if` on line 130...

File: src/main/java/sk/freemap/gpxAnimator/ui/GeneralSettingsPanel.java [hunk 12]
```diff
@@ -89,52 +91,52 @@ public GeneralSettingsPanel() {
...
+		outputFileSelector.setToolTipText(Option.OUTPUT.getHelp());
+		final GridBagConstraints gbc_outputFileSelector = new GridBagConstraints();
+		gbc_outputFileSelector.fill = GridBagConstraints.BOTH;
+		gbc_outputFileSelector.insets = new Insets(0, 0, 5, 0);
+		gbc_outputFileSelector.gridx = 1;
+		gbc_outputFileSelector.gridy = 0;
+		add(outputFileSelector, gbc_outputFileSelector);
 
-				outputFileSelector.addPropertyChangeListener("filename", propertyChangeListener);
+		outputFileSelector.addPropertyChangeListener("filename", propertyChangeListener);
```
Comment (McPringle) [inline src/main/java/sk/freemap/gpxAnimator/ui/GeneralSettingsPanel.java]: There are so many code formatting changes in this file, I really have problems identifying your code changes in these parts! Please revert all your code formatting changes, so your PR only contains the changed code. If you want to change the code formatting style, we can discuss it and do it in a separate commit and PR.

File: src/main/java/sk/freemap/gpxAnimator/ui/GeneralSettingsPanel.java [hunk 13]
```diff
@@ -666,6 +668,54 @@ protected Type configure(final JFileChooser outputFileChooser) {
...
+		final GridBagConstraints gbc_tileCachePathSelector = new GridBagConstraints();
+		gbc_tileCachePathSelector.fill = GridBagConstraints.BOTH;
+		gbc_tileCachePathSelector.insets = new Insets(0, 0, 5, 0);
+		gbc_tileCachePathSelector.gridx = 1;
+		gbc_tileCachePathSelector.gridy = 22;
+		add(tileCachePathSelector, gbc_tileCachePathSelector);
+
+		tileCachePathSelector.addPropertyChangeListener("filename", propertyChangeListener);
+
+		final JLabel lblTileCacheTimeLimit = new JLabel("Tile cache age limit");
```
Comment (McPringle) [inline src/main/java/sk/freemap/gpxAnimator/ui/GeneralSettingsPanel.java]: Tile cache *time* limit?

File: src/main/java/sk/freemap/gpxAnimator/ui/GeneralSettingsPanel.java [hunk 14]
```diff
@@ -745,6 +795,8 @@ public void setConfiguration(final Configuration c) {
 		totalTimeSpinner.setValue(c.getTotalTime());
 		backgroundMapVisibilitySlider.setValue((int) (c.getBackgroundMapVisibility() * 100));
 		photoTimeSpinner.setValue(c.getPhotoTime());
+		tileCacheTimeLimitSpinner.setValue(1000L*c.getTileCacheTimeLimit());
```
Comment (McPringle) [inline src/main/java/sk/freemap/gpxAnimator/ui/GeneralSettingsPanel.java]: Seconds to milliseconds? A short hint as a comment would be nice.

**Comment by n76** [comment]
I think I've addressed your concerns. I made some other changes to the tile caching object that tie in indirectly. First I always use the uncached download method so the user agent, etc. are all in one place and used consistently. Secondly, I try to treat a cache failure by falling back to uncached behavior. Finally, I decided to add a file extension to the cached tiles so they can easily be opened by image handling apps by double clicking on them.

Looking at my cache directory, I think I need to handle the aging differently. At present I only age the files we are looking at. This means that if/when you finish making a video the cached tiles are likely to remain for a long time (until you make another animation that happens to use those same tiles).

**Comment by McPringle** [comment]
This PR closes issue #77.
